2018 Robot/src/org/usfirst/frc/team4999/utils/PIDFactory.java

```diff
@@ -0,0 +1,393 @@
+	
+	private static final PIDThread calculator = new PIDThread(7);
+	
+	public static void addToCalculator(MomentumPID cont) {
+		calculator.addController(cont);
+		if(!calculator.isAlive() && !calculator.isDead())
+			calculator.start();
+	}
+	
+	private static Properties openFile(String file) {
+		System.out.println("Loading " + file);
+		Properties ret = new Properties();
+		FileInputStream input = null;
+		try {
+			input = new FileInputStream(file);
+			ret.load(input);
+		} catch (IOException ex) {
+			ex.printStackTrace();
+		} finally {
+			if(input != null) {
+				try {
+					input.close();
+				} catch(IOException e) {
+					e.printStackTrace();
+				}
+			}
+		}
+		return ret;
+	}
+	
+	private static void saveFile(Properties props, String file) {
+		System.out.println("Saving " + file);
+		FileOutputStream output = null;
+		try {
+			output = new FileOutputStream(file);
+			props.store(output, null);
+		} catch (IOException ex) {
+			ex.printStackTrace();
+		} finally {
+			if(output != null) {
+				try {
+					output.close();
+				} catch(IOException e) {
+					e.printStackTrace();
+				}
+			}
+		}
+	}
+	
+	private static void savePID(MomentumPID pid, String path) {
+		Properties props = new Properties();
+		props.put("P", String.format("%f", pid.getP()));
+		props.put("I", String.format("%f", pid.getI()));
+		props.put("D", String.format("%f", pid.getD()));
+		props.put("Error Zone", String.format("%f", pid.getErrZone()));
+		props.put("Target Zone", String.format("%f", pid.getTargetZone()));
+		props.put("Target Time", String.format("%f", pid.getTargetTime()));
+		saveFile(props, path);
+	}
+	
+	private static boolean checkFile(String file) {
+		File f = new File(file);
+		return f.exists() && !f.isDirectory();
+	}
+	
+	static class AverageEncoder implements PIDSource{
+		private Encoder left, right;
+		
+		private static final int SAMPLES = 10;
+
+		private PIDSourceType sourcetype = PIDSourceType.kDisplacement;
+
+
+		
+		public AverageEncoder(Encoder left, Encoder right) {
+			this.left = left;
+			this.right = right;
+		}
+		
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			sourcetype = pidSource;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return sourcetype;
+		}
+
+		@Override
+		public double pidGet() {
+			switch(sourcetype) {
+			case kRate:
+				return (left.getRate() + right.getRate()) / 2;
+			case kDisplacement:
+			default:
+				return (left.getDistance() + right.getDistance()) / 2;
+			}
+		}
+		
+	}
+	
+	static class GyroPitch implements PIDSource {
+		PIDSourceType type = PIDSourceType.kDisplacement;
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			type = pidSource;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return type;
+		}
+
+		@Override
+		public double pidGet() {
+			double angle = RobotMap.gyro.getPitch();
+			if(RobotMap.gyro.currentSensor() == Sensor.ADIS)
```

Comment:
Do we not trust the ADIS for pitch angle?

Reply:
No, the AHRS filter doesn't work correctly and the raw gyro drifts

Reply:
Then why do we trust it at all?

Reply:
It's better than nothing for rotation about the z-axis